datalad_next/runners/run.py

```diff
@@ -0,0 +1,64 @@
+"""
+"""
+
+from __future__ import annotations
+
+from contextlib import contextmanager
+from pathlib import Path
+from queue import Queue
+from subprocess import (
+    DEVNULL,
+    TimeoutExpired,
+)
+from typing import (
+    IO,
+)
+
+from datalad.runner.nonasyncrunner import _ResultGenerator
+
+from . import (
+    Protocol,
+    ThreadedRunner,
+)
+
+
+@contextmanager
+def run(
+    cmd: list,
+    protocol_class: Protocol,
+    *,
+    cwd: Path | None = None,
+    input: int | IO | bytes | Queue[bytes | None] | None = None,
+    # only generator protocols make sense for timeout, and timeouts are
+    # only checked when the generator polls
+    timeout: float | None = None,
+) -> dict | _ResultGenerator:
+    runner = ThreadedRunner(
+        cmd=cmd,
+        protocol_class=protocol_class,
+        stdin=DEVNULL if input is None else input,
+        cwd=cwd,
+        timeout=timeout,
+    )
+    try:
+        yield runner.run()
+    finally:
+        # if we get here the subprocess has no business running
+        # anymore. When run() exited normally, this should
+        # already be the case -- we make sure that no zombies
+        # accumulate
+        if runner.process is not None:
```

Comment:
Generally, the result generator should get a chance to pull all data from the subprocess by calling something like `tuple(result_generator)`. Threaded runner has support for timeouts and that includes processes. The runner will 
`terminate`, but not `kill`, a process if the timeout handler returns `True` for a process timeout (that is a timeout where the `fd` is `None).  If the process triggers multiple timeouts, the timeout-callback could theoretically kill it. 

Having said that, the code here should work, as far as I can see. And it actually introduces "killing", which the ThreadedRunner does not automatically invoke (that would have to be done as a result of timeout-calls)

Reply:
It does not. I tried to use it for the SSH uploader implementation in here, and it fails to upload stuff.